pkg/tfgen/docs.go

```diff
@@ -867,69 +876,259 @@ func getNestedBlockName(line string) string {
+			return ast.WalkContinue
+		}
+
+		var key docsPath
+		if child != nil && parent != child {
+			if child.writable {
+				key = docsPath(parent.name).join(child.name)
+			}
+		} else if latest != nil && parent == nil {
+			if latest.writable {
+				key = docsPath(latestParent.name).join(latest.name)
 			}
-			ret.Arguments[docsPath(lastMatch)].description += line
+		} else if parent != nil {
+			if parent.writable {
+				key = docsPath(parent.name)
+			}
+		} else if latestParent != nil {
+			if latestParent.writable {
+				key = docsPath(latestParent.name)
+			}
+		}
+
+		if key == "" {
+			return ast.WalkContinue
+		}
+
+		desc := string(renderMdNode(src, node))
+
+		arg := ret.Arguments[key]
+		if arg == nil {
+			arg = &argumentDocs{}
+			ret.Arguments[key] = arg
 		}
+		arg.description += desc
+		return ast.WalkContinue
 	}
 
-	var hadSpace bool
-	for _, line := range subsection {
-		if name, desc, matchFound := parseArgFromMarkdownLine(line); matchFound {
-			// We have found a new
-			addNewHeading(name, desc, line)
-			lastMatch = name
-		} else if strings.TrimSpace(line) == "---" {
-			// --- is a markdown section break. This probably indicates the
-			// section is over, but we take it to mean that the current
-			// heading is over.
-			lastMatch = ""
-		} else if nestedBlockCurrentLine := getNestedBlockName(line); hadSpace && nestedBlockCurrentLine != "" {
-			nested = docsPath(nestedBlockCurrentLine)
-			lastMatch = ""
-		} else if !isBlank(line) && lastMatch != "" {
-			extendExistingHeading(line)
-		} else if nestedBlockCurrentLine := getNestedBlockName(line); nestedBlockCurrentLine != "" {
-			nested = docsPath(nestedBlockCurrentLine)
-			lastMatch = ""
-		} else if lastMatch != "" {
-			extendExistingHeading(line)
+	// Identify the transition points between arguments.
+	err := ast.Walk(node, func(node ast.Node, entering bool) (ast.WalkStatus, error) {
+		if entering {
+			return descend(node), nil
+		}
+
+		if node.Kind() == ast.KindList {
+			// When we exit a list, clear our current state
+			if !entering {
+				parent = nil
+				child = nil
+			}
+			return ast.WalkContinue, nil
 		}
-		hadSpace = isBlank(line)
+
+		defer func() {
+			if child != nil && child.node == node {
+				child = nil
+			} else if parent != nil && parent.node == node {
+				parent = nil
+			}
+		}()
+
+		// Write the state on exit:
+
+		// There are several early exit conditions:
+		if (latest == nil && latestParent == nil) || // Don't have a node to write to
+			node == latestParent.node && latestParent == child ||
+			skipPrint[node] { // Is a marker node
+			return ast.WalkContinue, nil
+		}
+
+		return write(node), nil
+	})
+	contract.AssertNoErrorf(err, "An error is never returned")
+
+	for k, v := range ret.Arguments {
+		v.description = cleanDescription(k, v.description)
 	}
+}
 
-	for _, v := range ret.Arguments {
-		v.description = strings.TrimRightFunc(v.description, unicode.IsSpace)
+func renderMdNode(src []byte, n ast.Node) []byte {
```

Comment:
I've run into this before; so I can guess. why are we rendering back? Trying to run regexes on a subtree? 

I think I had to do something like this before an it was very disappointing that markdown library didn't provide a full featured AST->markdown renderer.

Reply:
Basically, the MD parser doesn't provide clean spans back into the original text. Having walked the parsed nodes to find what we want to extract, we need to render back to text the selected bit.

If it was possible, I would like this function to read `return src[n.Span().Start:n.Span().End]`, but that isn't supported.